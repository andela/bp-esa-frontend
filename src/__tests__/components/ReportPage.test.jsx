/* global mount */
import React from 'react';
import configureStore from 'redux-mock-store';
import ReportComponent, { ReportPage, mapDispatchToProps, mapStateToProps } from '../../components/ReportPage';
import { sampleReports, stats } from '../../fixtures/fixtures';

const state = {
  automation: {
    data: sampleReports.data,
    error: { error: '' },
    isLoading: false,
    pagination: sampleReports.pagination,
  },
  stats,
  error: {},
  isLoading: false,
};

const mockStore = configureStore();
const store = mockStore(state);

describe('ReportPage Component', () => {
  const props = {
    automation: {
      isLoading: false,
      data: sampleReports.data,
      pagination: sampleReports.pagination,
      error: {},
    },
    stats,
    location: { search: '?view=cardView' },
    currentUser: {
      additionalUserInfo: {
        profile: {
          name: '',
          picture: '',
        },
      },
    },
    history: { push: jest.fn() },
    removeCurrentUser: jest.fn(),
    formatDates: jest.fn(),
    fetchAllAutomation: jest.fn(() => Promise.resolve()),
    fetchUpdates: jest.fn(),
    resetUpdates: jest.fn(),
    fetchStat: jest.fn(),
    realTimeReport: sampleReports.data,
    retryFailedAutomation: jest.fn,
  };
  let component;

  beforeEach(() => {
    component = mount(<ReportPage {...props} />);
  });

  afterEach(() => {
    component.unmount();
  })

  it('should render as expected', () => {
    const title = component.find('.text');
    expect(title.text()).toEqual('ESA');
  });

  it('should redirect to the AIS page when you click the engineer\'s name', () => {
    component.setState({ viewMode: 'listView' });
    const redirectToAIS = component.find('.table-body')
      .find('.report-table')
      .find('tbody')
      .find('tr')
      .find('.fellow')
      .at(0);

    global.open = jest.fn();
    redirectToAIS.simulate('click');
    expect(global.open).toHaveBeenCalled();
  });

  it('updates the page limit', () => {
    const { pagination, filters } = component.state();
    const limitSelection = component.find('.select');
    limitSelection.simulate('change', { target: { value: 25 } });
    expect(component.state('pagination').limit).toEqual(25);
    expect(props.fetchAllAutomation).toHaveBeenCalledWith(pagination, filters);
  });

  it('updates the current page state when user types in the input box', () => {
    const currentPageInput = component.find('.form-input');
    currentPageInput.simulate('change', { target: { value: 30 } });
    expect(component.state('tempCurrentPage')).toEqual(30);
  });

  describe('render view', () => {
    let wrapper;
    it('should render view of listCard', () => {
      component.setState({ viewMode: 'listView' });
      const tableBody = component.find('.table-body');
      expect(tableBody.html()).toContain('onboarding');
      component.setState({ reportData: sampleReports.data, isLoadingReports: false });

      const infoButton = component
        .find('.fa-info-circle')
        .at(0);

      infoButton.simulate('click');
      component.find('.modal-close-button-group').simulate('click');
      expect(component.instance().state.isModalOpen).toEqual(false);
    });

    describe('render view', () => {
      it('should render card view', () => {
        const viewButton = component
          .find('.report-navbar-container')
          .find('.report-navbar')
          .find('#card-icon');
        viewButton.simulate('click');
        expect(component.state().viewMode).toEqual('cardView');
      });

      it('should render list', () => {
        const viewButton = component
          .find('.report-navbar-container')
          .find('.report-navbar')
          .find('#list-icon');
        viewButton.simulate('click');
        expect(component.state().viewMode).toEqual('listView');
      });
    });

    it('should call handleRetryAutomation for the card view', () => {
      wrapper = mount(<ReportPage {...props} />);
      const instance = wrapper.instance();
      jest.spyOn(instance, 'handleRetryAutomation');
      const button = wrapper.find('#retry-automation');
      button.simulate('click');
      // eslint-disable-next-line no-unused-expressions
      expect(instance.handleRetryAutomation).toBeCalled;
    });

    it('should call handleRetryAutomation for the table modal view', () => {
      component.setState({ viewMode: 'listView' });
      component.setState({ reportData: sampleReports, isLoadingReports: false });
      const instance = component.instance();
      jest.spyOn(instance, 'handleRetryAutomation');
      component.find('#info-icon')
        .at(0)
        .simulate('click');
      const button = component.find('.retry-btn');
      button.simulate('click');
      // eslint-disable-next-line no-unused-expressions
      expect(instance.handleRetryAutomation).toBeCalled;
    });
  });

  describe('The mapStateToProps', () => {
    it('should return the expected props object', () => {
      const storeState = {
        stats: {
          isLoading: false,
          data: {},
          error: {},
        },
      };

      const expectedProps = mapStateToProps(storeState);
      expect(expectedProps.stats).toEqual(storeState.stats);
    });
  });

  describe('The mapDispatchToProps', () => {
    let dispatch;
    let expectedProps;

    beforeEach(() => {
      dispatch = jest.fn();
      expectedProps = mapDispatchToProps(dispatch);
    });

    it('should ensure that fetchStat is mapped to props', () => {
      expectedProps.fetchStat();
      expect(dispatch).toHaveBeenCalled();
    });

    it('should ensure that fetchAllAutomation is mapped to props', () => {
      expectedProps.fetchAllAutomation();
      expect(dispatch).toHaveBeenCalled();
    });

    it('should ensure that fetchUpdates is mapped to props', () => {
      expectedProps.fetchUpdates();
      expect(dispatch).toHaveBeenCalled();
    });

    it('should ensure that resetUpdates is mapped to props', () => {
      expectedProps.resetUpdates();
      expect(dispatch).toHaveBeenCalled();
    });

    it('should ensure that fetchStat is mapped to props', () => {
      expectedProps.fetchStat();
      expect(dispatch).toHaveBeenCalled();
    });
  });

  describe('Test component methods', () => {
    it('should handle update tab', () => {
      const updateTab = component.find('.update-tab');
      updateTab.simulate('click');
      expect(props.resetUpdates).toHaveBeenCalled();
    });

    it('should handle page navigation', () => {
      const nextPageBtn = component.find('.page-btn').first();
      nextPageBtn.simulate('click');
      expect(props.fetchAllAutomation).toHaveBeenCalled();
      const tableBody = component.find('.table-body').find('.report-table');
      expect(tableBody).toBeDefined();
    });
  });
});
