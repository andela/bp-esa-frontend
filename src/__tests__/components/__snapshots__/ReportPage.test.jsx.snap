// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportPage /> should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReportPage
    currentUser={
      Object {
        "additionalUserInfo": Object {
          "profile": Object {
            "name": "",
            "picture": "",
          },
        },
      }
    }
    formatDates={[MockFunction]}
    history={Object {}}
    removeCurrentUser={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          currentUser={
            Object {
              "additionalUserInfo": Object {
                "profile": Object {
                  "name": "",
                  "picture": "",
                },
              },
            }
          }
          history={Object {}}
          removeCurrentUser={[MockFunction]}
        />,
        <ReportNavBar
          filter={[Function]}
        />,
        <div
          id="report-page"
        >
          <div
            className="table-header"
          >
            <table
              className="report-table"
            >
              <thead>
                <tr>
                  <th
                    className="numbering"
                  >
                    #
                  </th>
                  <th>
                    Date
                  </th>
                  <th>
                    Fellow Name
                  </th>
                  <th>
                    Partner Name
                  </th>
                  <th>
                    Type
                  </th>
                  <th>
                    Slack
                  </th>
                  <th>
                    Email
                  </th>
                  <th>
                    Freckle
                  </th>
                </tr>
              </thead>
            </table>
          </div>
          <div
            className="table-body"
          >
            <Spinner
              size="large"
            />
          </div>
          <AutomationDetails
            closeModal={[Function]}
            formatDates={[Function]}
            isModalOpen={false}
            modalContent={Object {}}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "currentUser": Object {
            "additionalUserInfo": Object {
              "profile": Object {
                "name": "",
                "picture": "",
              },
            },
          },
          "history": Object {},
          "removeCurrentUser": [MockFunction],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "filter": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="table-header"
            >
              <table
                className="report-table"
              >
                <thead>
                  <tr>
                    <th
                      className="numbering"
                    >
                      #
                    </th>
                    <th>
                      Date
                    </th>
                    <th>
                      Fellow Name
                    </th>
                    <th>
                      Partner Name
                    </th>
                    <th>
                      Type
                    </th>
                    <th>
                      Slack
                    </th>
                    <th>
                      Email
                    </th>
                    <th>
                      Freckle
                    </th>
                  </tr>
                </thead>
              </table>
            </div>,
            <div
              className="table-body"
            >
              <Spinner
                size="large"
              />
            </div>,
            <AutomationDetails
              closeModal={[Function]}
              formatDates={[Function]}
              isModalOpen={false}
              modalContent={Object {}}
            />,
          ],
          "id": "report-page",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <table
                className="report-table"
              >
                <thead>
                  <tr>
                    <th
                      className="numbering"
                    >
                      #
                    </th>
                    <th>
                      Date
                    </th>
                    <th>
                      Fellow Name
                    </th>
                    <th>
                      Partner Name
                    </th>
                    <th>
                      Type
                    </th>
                    <th>
                      Slack
                    </th>
                    <th>
                      Email
                    </th>
                    <th>
                      Freckle
                    </th>
                  </tr>
                </thead>
              </table>,
              "className": "table-header",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <thead>
                  <tr>
                    <th
                      className="numbering"
                    >
                      #
                    </th>
                    <th>
                      Date
                    </th>
                    <th>
                      Fellow Name
                    </th>
                    <th>
                      Partner Name
                    </th>
                    <th>
                      Type
                    </th>
                    <th>
                      Slack
                    </th>
                    <th>
                      Email
                    </th>
                    <th>
                      Freckle
                    </th>
                  </tr>
                </thead>,
                "className": "report-table",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <tr>
                    <th
                      className="numbering"
                    >
                      #
                    </th>
                    <th>
                      Date
                    </th>
                    <th>
                      Fellow Name
                    </th>
                    <th>
                      Partner Name
                    </th>
                    <th>
                      Type
                    </th>
                    <th>
                      Slack
                    </th>
                    <th>
                      Email
                    </th>
                    <th>
                      Freckle
                    </th>
                  </tr>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <th
                        className="numbering"
                      >
                        #
                      </th>,
                      <th>
                        Date
                      </th>,
                      <th>
                        Fellow Name
                      </th>,
                      <th>
                        Partner Name
                      </th>,
                      <th>
                        Type
                      </th>,
                      <th>
                        Slack
                      </th>,
                      <th>
                        Email
                      </th>,
                      <th>
                        Freckle
                      </th>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "#",
                        "className": "numbering",
                      },
                      "ref": null,
                      "rendered": "#",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Date",
                      },
                      "ref": null,
                      "rendered": "Date",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Fellow Name",
                      },
                      "ref": null,
                      "rendered": "Fellow Name",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Partner Name",
                      },
                      "ref": null,
                      "rendered": "Partner Name",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Type",
                      },
                      "ref": null,
                      "rendered": "Type",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Slack",
                      },
                      "ref": null,
                      "rendered": "Slack",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Email",
                      },
                      "ref": null,
                      "rendered": "Email",
                      "type": "th",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Freckle",
                      },
                      "ref": null,
                      "rendered": "Freckle",
                      "type": "th",
                    },
                  ],
                  "type": "tr",
                },
                "type": "thead",
              },
              "type": "table",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Spinner
                size="large"
              />,
              "className": "table-body",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "size": "large",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "closeModal": [Function],
              "formatDates": [Function],
              "isModalOpen": false,
              "modalContent": Object {},
              "modalType": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            currentUser={
              Object {
                "additionalUserInfo": Object {
                  "profile": Object {
                    "name": "",
                    "picture": "",
                  },
                },
              }
            }
            history={Object {}}
            removeCurrentUser={[MockFunction]}
          />,
          <ReportNavBar
            filter={[Function]}
          />,
          <div
            id="report-page"
          >
            <div
              className="table-header"
            >
              <table
                className="report-table"
              >
                <thead>
                  <tr>
                    <th
                      className="numbering"
                    >
                      #
                    </th>
                    <th>
                      Date
                    </th>
                    <th>
                      Fellow Name
                    </th>
                    <th>
                      Partner Name
                    </th>
                    <th>
                      Type
                    </th>
                    <th>
                      Slack
                    </th>
                    <th>
                      Email
                    </th>
                    <th>
                      Freckle
                    </th>
                  </tr>
                </thead>
              </table>
            </div>
            <div
              className="table-body"
            >
              <Spinner
                size="large"
              />
            </div>
            <AutomationDetails
              closeModal={[Function]}
              formatDates={[Function]}
              isModalOpen={false}
              modalContent={Object {}}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "currentUser": Object {
              "additionalUserInfo": Object {
                "profile": Object {
                  "name": "",
                  "picture": "",
                },
              },
            },
            "history": Object {},
            "removeCurrentUser": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "filter": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="table-header"
              >
                <table
                  className="report-table"
                >
                  <thead>
                    <tr>
                      <th
                        className="numbering"
                      >
                        #
                      </th>
                      <th>
                        Date
                      </th>
                      <th>
                        Fellow Name
                      </th>
                      <th>
                        Partner Name
                      </th>
                      <th>
                        Type
                      </th>
                      <th>
                        Slack
                      </th>
                      <th>
                        Email
                      </th>
                      <th>
                        Freckle
                      </th>
                    </tr>
                  </thead>
                </table>
              </div>,
              <div
                className="table-body"
              >
                <Spinner
                  size="large"
                />
              </div>,
              <AutomationDetails
                closeModal={[Function]}
                formatDates={[Function]}
                isModalOpen={false}
                modalContent={Object {}}
              />,
            ],
            "id": "report-page",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <table
                  className="report-table"
                >
                  <thead>
                    <tr>
                      <th
                        className="numbering"
                      >
                        #
                      </th>
                      <th>
                        Date
                      </th>
                      <th>
                        Fellow Name
                      </th>
                      <th>
                        Partner Name
                      </th>
                      <th>
                        Type
                      </th>
                      <th>
                        Slack
                      </th>
                      <th>
                        Email
                      </th>
                      <th>
                        Freckle
                      </th>
                    </tr>
                  </thead>
                </table>,
                "className": "table-header",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <thead>
                    <tr>
                      <th
                        className="numbering"
                      >
                        #
                      </th>
                      <th>
                        Date
                      </th>
                      <th>
                        Fellow Name
                      </th>
                      <th>
                        Partner Name
                      </th>
                      <th>
                        Type
                      </th>
                      <th>
                        Slack
                      </th>
                      <th>
                        Email
                      </th>
                      <th>
                        Freckle
                      </th>
                    </tr>
                  </thead>,
                  "className": "report-table",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <tr>
                      <th
                        className="numbering"
                      >
                        #
                      </th>
                      <th>
                        Date
                      </th>
                      <th>
                        Fellow Name
                      </th>
                      <th>
                        Partner Name
                      </th>
                      <th>
                        Type
                      </th>
                      <th>
                        Slack
                      </th>
                      <th>
                        Email
                      </th>
                      <th>
                        Freckle
                      </th>
                    </tr>,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <th
                          className="numbering"
                        >
                          #
                        </th>,
                        <th>
                          Date
                        </th>,
                        <th>
                          Fellow Name
                        </th>,
                        <th>
                          Partner Name
                        </th>,
                        <th>
                          Type
                        </th>,
                        <th>
                          Slack
                        </th>,
                        <th>
                          Email
                        </th>,
                        <th>
                          Freckle
                        </th>,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "#",
                          "className": "numbering",
                        },
                        "ref": null,
                        "rendered": "#",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Date",
                        },
                        "ref": null,
                        "rendered": "Date",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Fellow Name",
                        },
                        "ref": null,
                        "rendered": "Fellow Name",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Partner Name",
                        },
                        "ref": null,
                        "rendered": "Partner Name",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Type",
                        },
                        "ref": null,
                        "rendered": "Type",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Slack",
                        },
                        "ref": null,
                        "rendered": "Slack",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Email",
                        },
                        "ref": null,
                        "rendered": "Email",
                        "type": "th",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Freckle",
                        },
                        "ref": null,
                        "rendered": "Freckle",
                        "type": "th",
                      },
                    ],
                    "type": "tr",
                  },
                  "type": "thead",
                },
                "type": "table",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Spinner
                  size="large"
                />,
                "className": "table-body",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": "large",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "closeModal": [Function],
                "formatDates": [Function],
                "isModalOpen": false,
                "modalContent": Object {},
                "modalType": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
